### PR TITLE
Let first frame listeners self remove

### PR DESCRIPTION
**Pull request by xster**
See test:

Let the listener callbacks remove themselves for a one-time callback without java.util.ConcurrentModificationException.

### CODE REVIEW

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    // The callback removed itself from the listener list. A second call doesn't call the callback.
+    jniUnderTest.onFirstFrame();
+    assertEquals(1, callbackCalled.get());
+  }
+
+  private static class FlutterJNITest extends FlutterJNI {
+    protected void onFirstFrame() {
+      // Exposed here for tests.
+      super.onFirstFrame();
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: Why is this here?

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+@Config(manifest=Config.NONE)
+@RunWith(RobolectricTestRunner.class)
+public class FlutterRendererTest {
+  FlutterJNITest jniUnderTest;
+  FlutterRenderer rendererUnderTest;
+
+  @Before
+  public void setUp() {
+    jniUnderTest = new FlutterJNITest();
+    rendererUnderTest = new FlutterRenderer(jniUnderTest);
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: I recommend instantiating the object under test within each test. You never know when a test is going to be introduced that requires some kind of unusual instantiation, and when it is, that developer will be forced to change every other test to make that happen. Better to just let each test configure the object under test as it desired.

Comment (matthew-carroll) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: Also, I don't think we need to suffix things with "test"...everything in here is related to a test.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  @Test
+  public void firstFrameListenerCanRemoveThemselves() {
+    AtomicInteger callbackCalled = new AtomicInteger(0);
+    OnFirstFrameRenderedListener callback = new OnFirstFrameRenderedListener() {
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: I think you can probably accomplish this with a mock with less code and complexity than what you have here...

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  FlutterRenderer rendererUnderTest;
+
+  @Before
+  public void setUp() {
+    jniUnderTest = new FlutterJNITest();
+    rendererUnderTest = new FlutterRenderer(jniUnderTest);
+  }
+
+  @Test
+  public void firstFrameListenerCanRemoveThemselves() {
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: I'd like to follow a policy whereby any true unit test prefixes the test with "it", e.g., "itAllowsFirstFrameListenersToRemoveThemselves". This pattern will help ensure that unit tests really are unit tests by grammatically framing in terms of the object under test.

Comment (matthew-carroll) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: Also, this isn't really testing the broader capability of removing oneself as a listener. This is specifically testing the capabilityof removing oneself during a first frame callback. You may want to make that explicit in the test name.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    OnFirstFrameRenderedListener callback = new OnFirstFrameRenderedListener() {
+      @Override
+      public void onFirstFrameRendered() {
+        callbackCalled.incrementAndGet();
+        rendererUnderTest.removeOnFirstFrameRenderedListener(this);
+      };
+    };
+
+    rendererUnderTest.addOnFirstFrameRenderedListener(callback);
+    jniUnderTest.onFirstFrame();
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: I would recommend that in all tests we clearly demarcate between setup/execution/verification. In this case everything up to and including the `addOnFirstFrameRenderedListener` is part of the setup, the `onFirstFrame()` invocation is the execution, and everything that follows is verification.

In addition to placing a blank line and maybe a comment between these 2 lines above, you could also slightly refactor what comes below. I dont think you actually need to interleave 2 `onFirstFrame()`s and 2 `assert`s. You can simply invoke `onFirstFrame()` twice in a row, and then verify that the callback was invoked exactly 1 time.

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java [hunk 6]
```diff
@@ -275,15 +277,17 @@ public void removeOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedList
...
-  private void onFirstFrame() {
+  protected void onFirstFrame() {
     ensureRunningOnMainThread();
     if (renderSurface != null) {
       renderSurface.onFirstFrameRendered();
     }
     // TODO(mattcarroll): log dropped messages when in debug mode (https://github.com/flutter/flutter/issues/25391)
-
-    for (OnFirstFrameRenderedListener listener : firstFrameListeners) {
+    CopyOnWriteArrayList<OnFirstFrameRenderedListener> mutableListeners =
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java]: This leads me to believe that this is not a test of `FlutterRenderer` but rather a test of `FlutterJNI`. I think something needs to change in this PR to achieve correctness in that regard. If you're truly unit testing `FlutterRenderer` then you shouldn't have to make this protected. If you're testing `FlutterJNI` then the test suite should be switched accordingly.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Config(manifest=Config.NONE)
+@RunWith(RobolectricTestRunner.class)
```
Comment (matthew-carroll) [inline shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: Why does this need to be robolectric?

File: shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java (hunk 6 above)
Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java]: I think there's a couple of layers of testings we'll need either way. 

- I think we'll need strict unit tests. Where we're testing classes. I added a FlutterJNITest for this one. 
- Some of the classes are just not part of the public API are not going to be used. We should have medium tests that tests things between an API surface and some sort of abstraction layer. In this case, I think making things that platform_view_android_jni.cc calls also test callable probably makes sense. 
- We need integration tests that also includes C++ stuff but since those are heavier, they won't be per feature.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 7 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: FlutterRenderer imports android classes

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 2 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: - I think the first point was to not use test class setup/teardowns? It's probably a bit philosophical but I think it'll be efficient to just use industry standard practices (such as `cs/@before$ file:test.java`)
- I think this case is a bit exceptional but most of the time, there'd be things that are under test and a bunch more things that are mocks supporting those real classes under test. This makes it clear which are actors in the test and which aren't (e.g. `cs/"undertest =" file:test.java`) but I don't feel strongly about it.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 3 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: mockito is unfortunately a bit less capable for these level triggered stuff (mostly due to java)

An equivalent here will turn into 

```java
@Mock OnFirstFrameRenderedListener callback;
when(callback. onFirstFrameRendered()).thenAnswer(
    new Answer() {
         public Object answer(InvocationOnMock invocation) {
             rendererUnderTest.removeOnFirstFrameRenderedListener(callback);
         }
    });
```

which is even less directly readable.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 4 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: ok. Renamed. 

I think the only way to not remove themselves during the first frame would be something like itAllowsFirstFrameListenersToRemoveThemselvesWithoutUsingAHandlerToPostBackADeferredRemoval :D

How's itAllowsFirstFrameListenersToRemoveThemselvesInline?

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 1 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: Discussed offline. Probably won't need this class once I move the package up one level at which point it'll be in the same package as the FlutterJNI.

File: shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java (hunk 5 above)
Comment (xster) [reply shell/platform/android/test/io/flutter/embedding/engine/renderer/FlutterRendererTest.java]: I think it's still worth checking for the actual timing (e.g. that we don't have false positives on onFirstFrame only calling listeners on second invocation or something)

**Comment by xster** [comment]
cc @RedBrogdon since you wanted to follow along

**Comment by matthew-carroll** [comment]
I think we can close this PR due to related changes made here:
https://github.com/flutter/engine/pull/11357

**Comment by xster** [comment]
yup, thanks for re-doing all the tests
